pkg/clusterapi/aws/actuator.go

```diff
@@ -413,8 +464,100 @@ func (a *Actuator) Exists(cluster *clusterv1.Cluster, machine *clusterv1.Machine
 	if err != nil {
 		return false, fmt.Errorf("error getting EC2 client: %v", err)
 	}
+
+	instances, err := a.getRunningInstances(machine, client)
+	if err != nil {
+		return false, err
+	}
+	if len(instances) == 0 {
+		mLog.Debug("instance does not exist")
+		return false, nil
+	}
+
+	// If more than one result was returned, it will be handled in Update.
+	mLog.Debug("instance exists")
+	return true, nil
+}
+
+// updateStatus calculates the new machine status, checks if anything has changed, and updates if so.
+func (a *Actuator) updateStatus(machine *clusterv1.Machine, instance *ec2.Instance, mLog log.FieldLogger) error {
+
+	mLog.Debug("updating status")
+
+	// Starting with a fresh status as we assume full control of it here.
+	awsStatus := &cov1.AWSMachineProviderStatus{}
+
+	// Instance may have existed but been deleted outside our control, clear it's status if so:
+	if instance != nil {
+		awsStatus.InstanceID = instance.InstanceId
+		awsStatus.InstanceState = instance.State.Name
+		// Some of these pointers may still be nil (public IP and DNS):
+		awsStatus.PublicIP = instance.PublicIpAddress
+		awsStatus.PublicDNS = instance.PublicDnsName
+		awsStatus.PrivateIP = instance.PrivateIpAddress
+		awsStatus.PrivateDNS = instance.PrivateDnsName
+	}
+	mLog.Debug("finished calculating AWS status")
+
+	awsStatusRaw, err := controller.ClusterAPIMachineProviderStatusFromAWSMachineProviderStatus(awsStatus)
+	if err != nil {
+		mLog.Errorf("error encoding AWS provider status: %v", err)
+		return err
+	}
+
+	machineCopy := machine.DeepCopy()
+	machineCopy.Status.ProviderStatus = awsStatusRaw
+
+	if !equality.Semantic.DeepEqual(machine.Status, machineCopy.Status) {
+		mLog.Info("machine status has changed, updating")
+		machineCopy.Status.LastUpdated = metav1.Now()
+
+		_, err := a.clusterClient.ClusterV1alpha1().Machines(machineCopy.Namespace).UpdateStatus(machineCopy)
+		if err != nil {
+			mLog.Errorf("error updating machine status: %v", err)
+			return err
+		}
+	} else {
+		mLog.Debug("status unchanged")
+	}
+	return nil
+}
+
+func (a *Actuator) terminateInstances(client *ec2.EC2, instanceIDs []*string, mLog log.FieldLogger) error {
+	for _, instanceID := range instanceIDs {
+		mLog.WithField("instanceID", *instanceID).Info("terminating instance")
+	}
+
+	terminateInstancesRequest := &ec2.TerminateInstancesInput{
+		InstanceIds: instanceIDs,
+	}
+	_, err := client.TerminateInstances(terminateInstancesRequest)
```

Comment:
For a future update... I'm wondering if it makes sense to not return until we have verified that the provider instance has been actually updated to the terminated state. That way the machine API object can better reflect what's going on in the provider.

Reply:
Is there a delay between the request and the state change? That did not occur to me. I can add a todo.